### ISSUE TITLE
false positive unused_result when passing closure to macro

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#81086]
I tried this code:

```rust
#![warn(unused_results)]
#[macro_use]
macro_rules! run_function {
    ($func: ident) => {{
        &$func();
    }};
}

fn dummy() {}

pub fn main() {
    let fun = || dummy();
    run_function!(fun);
}

````
[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c4ba98093d3140e119f0776ebddd0eb3)

There's an unexpected warning about unused results although I don't see `Results` being used anywhere in the code.

Does rustc think that we could potentially pass a function that returns a result?
````
   Compiling playground v0.0.1 (/playground)
warning: unused result
  --> src/main.rs:5:9
   |
5  |         &$func();
   |         ^^^^^^^^^
...
13 |     run_function!(fun);
   |     ------------------- in this macro invocation
   |
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![warn(unused_results)]
   |         ^^^^^^^^^^^^^^
   = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

warning: 1 warning emitted
````

### Meta

`rustc --version --verbose`:
```
rustc 1.51.0-nightly (bc39d4d9c 2021-01-15)
binary: rustc
commit-hash: bc39d4d9c514e5fdb40a5782e6ca08924f979c35
commit-date: 2021-01-15
host: x86_64-unknown-linux-gnu
release: 1.51.0-nightly
LLVM version: 11.0
```

**Comment by PatchMixolydic** [comment]
This still happens without using macros or closures ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ab44f5bf57ea23f4523bfbe11d56ed92)):

```rust
#![warn(unused_results)]

fn dummy() {}

pub fn main() {
    &dummy();
}
```

```
warning: unused result
 --> src/main.rs:6:5
  |
6 |     &dummy();
  |     ^^^^^^^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(unused_results)]
  |         ^^^^^^^^^^^^^^
```

This only happens when taking a reference to the result of the function call. [unused_results] checks if the result of any expression is unused. Presumably, this lint ignores `()` (which is the type of `dummy()` and `fun()`), but not `&()` (which is the type of `&dummy()` and `&fun()`)!

I'm not entirely sure what the use case would be for taking a reference to the result of a function call and ignoring it, so the lint is probably correct.

To clarify the lint name, it's not that you're not using a `Result`, it's that you're not using the result of `&fun()`. To fix this, you have \[at least] two options:

* To make `run_function!` return a reference, remove the semicolon:
```rust
macro_rules! run_function {
    ($func: ident) => {{
        &$func()
    }};
}
```
* To silence the lint, remove the borrow:
```rust
macro_rules! run_function {
    ($func: ident) => {{
        $func();
    }};
}
```
  * If you're using this macro on functions that don't return `()`, you can place an `allow` attribute in front of the block (I think)
```rust
macro_rules! run_function {
    ($func: ident) => {
        #[allow(unused_results)]
        {
            &$func();
        }
    };
}
```

[unused_results]: https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#unused-results

**Comment by PatchMixolydic** [comment]
Additional note: `unused_results` isn't needed to see if an instance of the `Result` enum goes unused. `Result` is marked as [`must_use`], so it should be picked up by the [`unused_must_use`] lint, which is set to `warn` by default.

[`must_use`]: https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute
[`unused_must_use`]: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unused-must-use

**Comment by fmease** [comment]
Now the output is

```
warning: unused borrow that must be used
  --> src/main.rs:5:9
   |
5  |         &$func();
   |         ^^^^^^^^ the borrow produces a value
...
13 |     run_function!(fun);
   |     ------------------ in this macro invocation
   |
   = note: `#[warn(unused_must_use)]` on by default
   = note: this warning originates in the macro `run_function` (in Nightly builds, run with -Z macro-backtrace for more info)
help: use `let _ = ...` to ignore the resulting value
   |
5  |         let _ = &$func();
   |         +++++++
```

Which makes it clear that it's not about `Result`. Closing as completed. If you disagree I will ofc reopen it.
